pip/req/req_set.py

```diff
@@ -439,6 +461,11 @@ def _prepare_file(self, finder, req_to_install):
             # # vcs update or unpack archive # #
             # ################################ #
             if req_to_install.editable:
+                if require_hashes:
+                    raise InstallationError(
+                        'The editable requirement %s cannot be installed when '
+                        'requiring hashes, because there is no single file to '
+                        'hash.' % req_to_install)
```

Comment:
We could do some logic to hash an arbitrary directory. Something like:

```python
import hashlib

h = hashlib.sha256()
for f in sorted(listdir(dir)):
    h.update(f)
    h.update(open(f).read())
```

But it'll be more involved in that. I poked some smart folks and one mentioned:

```
[14:23:27]  <nwilcox>	dstufft: That is an ambiguous encodeing. Consider a directory with file "foo" containing "cat" and "bar" containing "dog", versus another with file "bar" containing "dogfoocat".
[14:23:51]  <nwilcox>	-but assuming the encoding to the hash input avoids such injections...
[14:24:07]  <dstufft>	nwilcox: good point!
[14:24:11]  <nwilcox>	another concern I might have is file system name encoding.
[14:24:30]  <nwilcox>	(Does the user's locale on linux cause weirdness?  I'm not sure.)
[14:24:45]  <nwilcox>	-it also ignores other metadata like modes or timestamps.
[14:25:58]  <nwilcox>	Related to filename encoding is sort. eg: GNU sort uses some kind of locale-specific logic unless you explicitly tell it you want byte-wise lexical sorting.
[14:26:19]  <nwilcox>	I don't know what python does for unicode sorting.
```

It'll be a bit tricky to get right, but I think that it could be done.